### PR TITLE
use nanmean for aggregating loss

### PR DESCRIPTION
**Pull request by winglian**
# What does this PR do?

When post training using context parallelism, some processes may have their chunk of the sample input masked out leading to a NaN loss for that process. Using `nanmean` allows us to keep the real loss that isn't `NaN`

@SunMarc 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2054,7 +2101,7 @@ def _maybe_log_save_evaluate(
             logs: dict[str, float] = {}
 
             # all_gather + mean() to get average loss over all processes
-            tr_loss_scalar = nested_gather(tr_loss, self.args.parallel_mode).mean().item()
+            tr_loss_scalar = nested_gather(tr_loss, self.args.parallel_mode).nanmean().item()
```
Comment (SunMarc) [inline src/transformers/trainer.py:2104]: we have the following already. Not sure how we are still getting the nan 

```python
                if (
                    self.args.logging_nan_inf_filter
                    and not is_torch_xla_available()
                    and (torch.isnan(tr_loss_step) or torch.isinf(tr_loss_step))
                ):
                    # if loss is nan or inf simply add the average of previous logged losses
                    self._tr_loss += self._tr_loss / (1 + self.state.global_step - self._globalstep_last_logged)
```

**Review by SunMarc — COMMENTED** [review]
Thanks, left a question

File: src/transformers/trainer.py (hunk 1 above)
Comment (winglian) [reply src/transformers/trainer.py:2104]: this was a fix we had to make last year when we added CP support, https://github.com/axolotl-ai-cloud/axolotl/pull/3033/changes, even though that code block above has been around for a couple of years. My assumption is that it has something to do with the multigpu gather.

Screenshot 2026-02-24 at 10 53 13 AM

Comment (SunMarc) [reply src/transformers/trainer.py:2104]: yeah most likely ! Thanks !

Comment (SunMarc) [inline src/transformers/trainer.py:2104]: can you add a check `self.args.logging_nan_inf_filter` if we are using `nanmean` ?

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2796,9 +2794,9 @@ def evaluation_loop(
         metrics = denumpify_detensorize(metrics)
 
         if isinstance(all_losses, list) and all_losses:
-            metrics[f"{metric_key_prefix}_loss"] = np.concatenate(all_losses).mean().item()
+            metrics[f"{metric_key_prefix}_loss"] = np.nanmean(np.concatenate(all_losses)).item()
         elif isinstance(all_losses, np.ndarray):
-            metrics[f"{metric_key_prefix}_loss"] = all_losses.mean().item()
+            metrics[f"{metric_key_prefix}_loss"] = np.nanmean(all_losses).item()
```
Comment (SunMarc) [inline src/transformers/trainer.py:2799]: same here

**Review by SunMarc — COMMENTED** [review]
Thanks, just a minor nit

**Review by qgallouedec — APPROVED** [review]
lgtm, cc @kashif
